### PR TITLE
More informative `Solver.__repr__`

### LINKED ISSUE #1926: Solver options printing
would be nice to be able to print all solver options to console/log for debugging purposes.

### PR DESCRIPTION
**Pull request by dweindl**
Closes #1926

Output of `print(amici.CVodeSolver())`:

```
<Swig Object of type 'amici::CVodeSolver *' at 0x7f988f70f090
  reporting_mode: <RDataReporting.full: 0>
  sens_meth: <SensitivityMethod.forward: 1>
  sens_order: <SensitivityOrder.none: 0>
  sens_meth_preeq: <SensitivityMethod.forward: 1>
  maxsteps: 10000
  maxtime: 1.7976931348623157e+308s
  abs_tol: 1e-16
  rel_tol: 1e-08
  abs_tol_b: 1e-16
  rel_tol_b: 1e-08
  abs_tol_fsa: 1e-16
  rel_tol_fsa: 1e-08
  abs_tol_quad: 1e-12
  rel_tol_quad: 1e-08
  abs_tol_ss: 1e-14
  rel_tol_ss: 1e-06
  abs_tol_sss: 1e-14
  rel_tol_sss: 1e-06
  int_sens_meth: <InternalSensitivityMethod.simultaneous: 1>
  int_type: <InterpolationType.hermite: 1>
  linsol: <LinearSolver.KLU: 9>
  lmm: <LinearMultistepMethod.BDF: 2>
  newton_damp_mode: <NewtonDampingFactorMode.on: 1>
  newton_damp_lb: 1e-08
  newton_maxsteps: 0
  newton_ss_check: False
  sens_ss_check: True
  interpolation_type: <InterpolationType.hermite: 1>
  ism: <InternalSensitivityMethod.simultaneous: 1>
  nlsol_iter: <NonlinearSolverIteration.newton: 2>
  stability_limit: True
  state_ordering: 0
>
```

### CODE REVIEW

File: swig/solver.i [hunk 1]
```diff
@@ -40,6 +40,50 @@ using namespace amici;
...
+%extend amici::CVodeSolver {
+%pythoncode %{
+def __repr__(self):
+    return "\n".join([
+        self.this.__repr__()[:-1],
+        "  reporting_mode: "
+        f"{RDataReporting(self.getReturnDataReportingMode())!r}",
+        f"  sens_meth: {SensitivityMethod(self.getSensitivityMethod())!r}",
+        f"  sens_order: {SensitivityOrder(self.getSensitivityOrder())!r}",
+        f"  sens_order: {SensitivityMethod(self.getSensitivityMethodPreequilibration())!r}",
```
Comment (dweindl) [inline swig/solver.i:53]:
```suggestion
        f"  sens_meth_preeq: {SensitivityMethod(self.getSensitivityMethodPreequilibration())!r}",
```

**Comment by FFroehlich** [comment]
> Closes #1926
> 
> Output of `print(amici.CVodeSolver())`:
> 
> ```
> <Swig Object of type 'amici::CVodeSolver *' at 0x7f988f70f090
>   reporting_mode: <RDataReporting.full: 0>
>   sens_meth: <SensitivityMethod.forward: 1>
>   sens_order: <SensitivityOrder.none: 0>
>   sens_meth_preeq: <SensitivityMethod.forward: 1>
>   maxsteps: 10000
>   maxtime: 1.7976931348623157e+308s
>   abs_tol: 1e-16
>   rel_tol: 1e-08
>   abs_tol_b: 1e-16
>   rel_tol_b: 1e-08
>   abs_tol_fsa: 1e-16
>   rel_tol_fsa: 1e-08
>   abs_tol_quad: 1e-12
>   rel_tol_quad: 1e-08
>   abs_tol_ss: 1e-14
>   rel_tol_ss: 1e-06
>   abs_tol_sss: 1e-14
>   rel_tol_sss: 1e-06
>   int_sens_meth: <InternalSensitivityMethod.simultaneous: 1>
>   int_type: <InterpolationType.hermite: 1>
>   linsol: <LinearSolver.KLU: 9>
>   lmm: <LinearMultistepMethod.BDF: 2>
>   newton_damp_mode: <NewtonDampingFactorMode.on: 1>
>   newton_damp_lb: 1e-08
>   newton_maxsteps: 0
>   newton_ss_check: False
>   sens_ss_check: True
>   interpolation_type: <InterpolationType.hermite: 1>
>   ism: <InternalSensitivityMethod.simultaneous: 1>
>   nlsol_iter: <NonlinearSolverIteration.newton: 2>
>   stability_limit: True
>   state_ordering: 0
> >
> ```

👍  neat, does this also work for SolverPtr out of the box?

**Review by FFroehlich — APPROVED** [review]

File: python/tests/test_swig_interface.py [hunk 2]
```diff
@@ -403,3 +403,7 @@ def test_model_instance_settings_custom_x0(pysb_example_presimulation_module):
     assert model2.getInitialStateSensitivities() == sx0
     assert settings == amici.get_model_settings(model2)
 
+
+def test_solver_repr_doesnt_raise():
+    print(amici.CVodeSolver())
+    print(repr(amici.CVodeSolver()))
```
Comment (FFroehlich) [inline python/tests/test_swig_interface.py:409]: isn't the default `__repr__` set by swig anyways (as for `self.this` anyway), so this wouldn't have raised even before these changes?

**Comment by dweindl** [comment]
> does this also work for SolverPtr out of the box?

Good point. Now it does.

> isn't the default __repr__ set by swig anyways (as for self.this anyway), so this wouldn't have raised even before these changes?

That's correct. It was mostly intended to check that no non-existing ReturnData fields are accessed. Now it also checks that it is not just the default `__repr__`.

**Pull request merged**